src/ui/ui_review_menu_nbgl.c

```diff
@@ -0,0 +1,332 @@
+    if (accept) {
+        if (txInfos.warnings[CUSTOM_CONTRACT_WARNING] == true) {
+            displayCustomContractWarning();
+        } else {
+            displayTransaction();
+        }
+    } else {
+        txInfos.rejectCb();
+        nbgl_useCaseStatus("Transaction rejected", false, ui_idle);
+    }
+}
+
+static void customContractWarningChoice(bool accept) {
+    if (accept) {
+        displayTransaction();
+    } else {
+        txInfos.rejectCb();
+        nbgl_useCaseStatus("Transaction rejected", false, ui_idle);
+    }
+}
+
+static void displayDataWarning(void) {
+    nbgl_useCaseChoice(&C_round_warning_64px,
+                       "WARNING\nData Present\nProceed with care",
+                       "Reject if you're not sure",
+                       "Continue",
+                       "Reject",
+                       dataWarningChoice);
+}
+
+static void displayCustomContractWarning(void) {
+    nbgl_useCaseChoice(&C_round_warning_64px,
+                       "WARNING\nCustom Contract\nProceed with care",
+                       "Reject if you're not sure",
+                       "Continue",
+                       "Reject",
+                       customContractWarningChoice);
+}
+
+static void displayTransaction(void) {
+    nbgl_useCaseStaticReview(&pairList, &infoLongPress, "Reject", reviewChoice);
+}
+
+static void reviewStart() {
+    nbgl_callback_t displayFunction = displayTransaction;
+    if (txInfos.warnings[DATA_WARNING] == true) {
+        displayFunction = displayDataWarning;
+    } else if (txInfos.warnings[CUSTOM_CONTRACT_WARNING] == true) {
+        displayFunction = displayCustomContractWarning;
+    }
+    nbgl_useCaseReviewStart(&C_stax_app_tron_64px,
+                            txInfos.flowTitle,
+                            txInfos.flowSubtitle,
+                            "Reject",
+                            displayFunction,
+                            rejectChoice);
+}
+
+static void reviewChoice(bool confirm) {
+    if (confirm) {
+        nbgl_useCaseStatus("TRANSACTION\nCONFIRMED", true, ui_idle);
+        txInfos.confirmCb();
+    } else {
+        rejectChoice();
+    }
+}
+
+static void rejectConfirmation(void) {
+    nbgl_useCaseStatus("Transaction rejected", false, ui_idle);
+    txInfos.rejectCb();
+}
+
+static void rejectChoice(void) {
+    nbgl_useCaseConfirm("Reject transaction?",
+                        NULL,
+                        "Yes, Reject",
+                        "Go back to transaction",
+                        rejectConfirmation);
+}
+
+static void prepareTxInfos(ui_approval_state_t state, bool data_warning) {
+    memset(&txInfos, 0, sizeof(txInfos));
+    memset(&infoLongPress, 0, sizeof(infoLongPress));
+
+    txInfos.warnings[DATA_WARNING] = data_warning;
+    txInfos.flowTitle = "Review Transaction";
+    txInfos.confirmCb = (nbgl_callback_t) io_seproxyhal_touch_tx_ok;
+    txInfos.rejectCb = (nbgl_callback_t) io_seproxyhal_touch_cancel;
```

Comment:
If common, maybe they don't need to be in `txInfos`

Reply:
@xchapron-ledger actually the callbacks can be different depending on the tx type : https://github.com/LedgerHQ/app-tron/blob/57f7a8ea36cdf6fb719ecacc5ba2d866dab02970/src/ui/ui_review_menu_nbgl.c#L313

So I use the structure to keep the correct one for later use.